### PR TITLE
Drive GetModificationTime using watched file events

### LINKED ISSUE #803: Replace GetModificationTime with a file version scheme via WatchedFile events
The idea is to replace `GetModificationTime`, which is expensive and called on every change, by a  new rule `GetFileVersion` implemented entirely on top of watched file events. This would reduce I/O, eliminate timestamp serialization and comparison, and hopefully improve performance 

One possible approach is to generalize the `FileExistsMapVar` in order to keep track of file versions.

### PR DESCRIPTION
**Pull request by pepeiborra**
This removes the need to poll the file system for source files (we still do it for interface files). 

The implementation is remarkably simple, as it uses the watch notifications only to *dirty* the Shake node, but we still reuse the existing code to compute the result of the rule.

Closes #803

### CODE REVIEW

**Comment by pepeiborra** [comment]
We need to change lsp-test to send `textDocument/didChange` notifications in `changeDoc`. @bubba @wz1000 any chances you could  make that change and upload a point release to Hackage?

**Comment by wz1000** [comment]
Doesn't it already send those notifications?

**Comment by pepeiborra** [comment]
> Doesn't it already send those notifications?

Ok, it does actually. The Haddocks are a lie...

**Comment by pepeiborra** [comment]
Ok, so the notification I had in mind is `‘workspace/didChangeWatchedFiles’` and it's only needed when changing non FOIs, which some tests do, in which case it's probably best to just sent it manually from the test

**Comment by wz1000** [comment]
BTW, I'm pretty sure vscode doesn't send file-changed notifications if the file is changed outside the editor. To support that case (other tools, git etc. can change files), I think it would be good to directly listen for file system events.

This would also let us handle iface files and similar.

`fsnotify`s interface seems easy enough to hook up into this: https://hackage.haskell.org/package/fsnotify-0.3.0.1/docs/System-FSNotify.html#v:watchTree

(There is also the slightly selfish reason, in that my editor doesn't send these events at all)

**Comment by pepeiborra** [comment]
It seems to be working for me. This is the message sent by VSCode on a git branch checkout:
```
2021-03-03 22:34:43.674318 [ThreadId 5] DEBUG haskell-lsp.parseOne:	---> {"jsonrpc":"2.0","method":"workspace/didChangeWatchedFiles","params":{"changes":[{"uri":"file:///Users/pepeiborra/scratch/ide/ghcide/test/data/plugin/KnownNat.hs","type":1},{"uri":"file:///Users/pepeiborra/scratch/ide/ghcide/test/data/plugin/RecordDot.hs","type":1},{"uri":"file:///Users/pepeiborra/scratch/ide/ghcide/src/Development/IDE/Core/FileExists.hs","type":2},{"uri":"file:///Users/pepeiborra/scratch/ide/ghcide/src/Development/IDE/Core/FileStore.hs","type":2},{"uri":"file:///Users/pepeiborra/scratch/ide/ghcide/src/Development/IDE/Core/Rules.hs","type":2},{"uri":"file:///Users/pepeiborra/scratch/ide/ghcide/src/Development/IDE/Core/Shake.hs","type":2},{"uri":"file:///Users/pepeiborra/scratch/ide/ghcide/src/Development/IDE/LSP/Notifications.hs","type":2},{"uri":"file:///Users/pepeiborra/scratch/ide/ghcide/test/exe/Main.hs","type":2}]}}
2021-03-03 22:34:43.674467 [ThreadId 27] DEBUG hls:	Watched file events: [FileEvent {_uri = Uri {getUri = "file:///Users/pepeiborra/scratch/ide/ghcide/test/data/plugin/KnownNat.hs"}, _xtype = FcCreated},FileEvent {_uri = Uri {getUri = "file:///Users/pepeiborra/scratch/ide/ghcide/test/data/plugin/RecordDot.hs"}, _xtype = FcCreated},FileEvent {_uri = Uri {getUri = "file:///Users/pepeiborra/scratch/ide/ghcide/src/Development/IDE/Core/FileExists.hs"}, _xtype = FcChanged},FileEvent {_uri = Uri {getUri = "file:///Users/pepeiborra/scratch/ide/ghcide/src/Development/IDE/Core/FileStore.hs"}, _xtype = FcChanged},FileEvent {_uri = Uri {getUri = "file:///Users/pepeiborra/scratch/ide/ghcide/src/Development/IDE/Core/Rules.hs"}, _xtype = FcChanged},FileEvent {_uri = Uri {getUri = "file:///Users/pepeiborra/scratch/ide/ghcide/src/Development/IDE/Core/Shake.hs"}, _xtype = FcChanged},FileEvent {_uri = Uri {getUri = "file:///Users/pepeiborra/scratch/ide/ghcide/src/Development/IDE/LSP/Notifications.hs"}, _xtype = FcChanged},FileEvent {_uri = Uri {getUri = "file:///Users/pepeiborra/scratch/ide/ghcide/test/exe/Main.hs"}, _xtype = FcChanged}]
2021-03-03 22:34:43.675346 [ThreadId 1912] DEBUG hls:	Finishing build session(exception: AsyncCancelled)
2021-03-03 22:34:43.694439 [ThreadId 27] DEBUG hls:	Restarting build session (aborting the previous one took 0.00s, profile saved at /tmp/ide-20210303-223215-00005.html)
2021-03-03 22:34:43.765524 [ThreadId 7] DEBUG haskell-lsp.sendServer:	<--2--{"jsonrpc":"2.0","params":{"value":{"kind":"report","message":"117/118"},"token":"1240"},"method":"$/progress"}
2021-03-03 22:34:45.928158 [ThreadId 5] DEBUG haskell-lsp.parseOne:	---> {"jsonrpc":"2.0","id":3,"method":"textDocument/codeLens","params":{"textDocument":{"uri":"file:///Users/pepeiborra/scratch/ide/ghcide/src/Development/IDE/LSP/Notifications.hs"}}}
2021-03-03 22:34:45.930564 [ThreadId 5] DEBUG haskell-lsp.parseOne:	---> {"jsonrpc":"2.0","id":4,"method":"textDocument/documentSymbol","params":{"textDocument":{"uri":"file:///Users/pepeiborra/scratch/ide/ghcide/src/Development/IDE/LSP/Notifications.hs"}}}
```

**Comment by pepeiborra** [comment]
BTW your client will continue to work with HLS, as long as it advertises its capabilities correctly. If it doesn't, we can add a flag

**Comment by pepeiborra** [comment]
Finally, I completely agree with the LSP spec on the topic of setting up our own file watches:

```
Servers are allowed to run their own file system watching mechanism and not rely on clients to provide file system events. However this is not recommended due to the following reasons:

to our experience getting file system watching on disk right is challenging, especially if it needs to be supported across multiple OSes.
file system watching is not for free especially if the implementation uses some sort of polling and keeps a file system tree in memory to compare time stamps (as for example some node modules do)
a client usually starts more than one server. If every server runs its own file system watching it can become a CPU or memory problem.
in general there are more server than client implementations. So this problem is better solved on the client side.
```

File: ghcide/src/Development/IDE/Core/FileStore.hs [hunk 1]
```diff
@@ -93,24 +96,39 @@ isFileOfInterestRule = defineEarlyCutoff $ \IsFileOfInterest f -> do
...
-              `catch` \(e :: IOException) -> do
-                let err | isDoesNotExistError e = "File does not exist: " ++ file'
-                        | otherwise = "IO error while reading " ++ file' ++ ", " ++ displayException e
-                    diag = ideErrorText file (T.pack err)
-                if isDoesNotExistError e && not missingFileDiags
-                    then return (Nothing, ([], Nothing))
-                    else return (Nothing, ([diag], Nothing))
+            Nothing -> do
+                isWF <- isWatched file
+                unless isWF alwaysRerun
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/FileStore.hs:111]: Does this kind of conditional-`alwaysRerun` work? I seem to recall getting very tangled up in this when touching the existence rules. I think it's "rule-global", so you can't actually do this, or you need to have two rules.

File: ghcide/src/Development/IDE/Core/FileStore.hs [hunk 2]
```diff
@@ -93,24 +96,39 @@ isFileOfInterestRule = defineEarlyCutoff $ \IsFileOfInterest f -> do
...
+                    `catch` \(e :: IOException) -> do
+                        let err | isDoesNotExistError e = "File does not exist: " ++ file'
+                                | otherwise = "IO error while reading " ++ file' ++ ", " ++ displayException e
+                            diag = ideErrorText file (T.pack err)
+                        if isDoesNotExistError e && not missingFileDiags
+                            then return (Nothing, ([], Nothing))
+                            else return (Nothing, ([diag], Nothing))
+
+-- | Reset the GetModificationTime state of watched files
+modifyFileStore :: IdeState -> [FileEvent] -> IO ()
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/FileStore.hs:122]: Should we call this `resetFileStore`, given that it doesn't actually use the event to modify the time? Or, indeed, could we get the time here and actually set the mod time?

File: ghcide/src/Development/IDE/Core/FileStore.hs (hunk 1 above)
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/FileStore.hs:111]: ping @ndmitchell

Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/FileStore.hs:114]: More evidence can be obtained easily via shake profiling and ghcide-bench. The following experiment:
```
cabal build ghcide-bench
mkdir /tmp/ghcide
cabal exec cabal run ghcide-bench -- -- --example-package-name Cabal --example-module Distribution/Simple.hs --example-package-version 3.0.0.0 -s "edit" --samples 1 --shake-profiling /tmp/ghcide --no-clean
```
Then open the latest html file in `/tmp/ghcide` and check the number of `GetModificationTime` rules executed in the last run.

BEFORE: 471
AFTER: 238 (for the `.hi` files)

image

**Comment by pepeiborra** [comment]
Can I get a stamp please?

File: ghcide/src/Development/IDE/Core/FileStore.hs (hunk 1 above)
Comment (michaelpj) [reply ghcide/src/Development/IDE/Core/FileStore.hs:114]: Thanks for checking, that looks convincing!

**Review by wz1000 — APPROVED** [review]

**Pull request merged**